src/dyninst/dyninst_comp.C

```diff
@@ -1426,123 +1426,116 @@ const char *fixUnderscores(const char *str)
-			if (func2 != NULL)
-				func2->getName(name2, name_max);
-
-			if ((func == NULL && func2 != NULL) ||
-					(func != NULL && func2 == NULL)) {
-				logerror("**Failed** test %d (%s)\n", test_num, test_name);
-				logerror("    frame->findFunction() disagrees with thread->findFunctionByEntry()\n");
-				logerror("    frame->findFunction() returns %s\n",
-						name);
-				logerror("    thread->findFunctionByEntry() return %s\n",
-						name2);
-				failed = true;
-				break;
-			} else if (func!=NULL && func2!=NULL && strcmp(name, name2)!=0) {
-				logerror("**Failed** test %d (%s)\n", test_num, test_name);
-				logerror("    BPatch_frame::findFunction disagrees with BPatch_thread::findFunctionByEntry\n");
-				failed = true;
-				break;
-			}
-
-			if (correct_frame_info[i].type != stack[j].getFrameType()) {
-				logerror("**Failed** test %d (%s)\n", test_num, test_name);
-				logerror("    Stack frame #%d has wrong type, is %s, should be %s\n", i+1, frameTypeString(stack[i].getFrameType()), frameTypeString(correct_frame_info[i].type));
-				logerror("    Stack frame 0x%lx, 0x%lx\n", stack[i].getPC(), stack[i].getFP() );
-				failed = true;
-				break;
-			}
+    // check if stack follows sequence in correct_frame_info
+    if(!isSubsequenceStack(stack, stack.size(), correct_frame_info, num_correct_names))
+    {
+        logerror("**Failed** test %d (%s)\n", test_num, test_name);
+        logerror("    Collected stack does not contain expected stack.\n");
+        failed = true;
+    }
 
-			if (stack[j].getFrameType() == BPatch_frameSignal ||
-					stack[j].getFrameType() == BPatch_frameTrampoline) {
-				// No further checking for these types right now
-			} else {
-				if (func == NULL) {
-					logerror("**Failed** test %d (%s)\n",
-							test_num, test_name);
-					logerror("    Stack frame #%d refers to an unknown function, should refer to %s\n", j+1, correct_frame_info[i].function_name);
-					failed = true;
-					break;
-				} else { /* func != NULL */
-					if (!hasExtraUnderscores(correct_frame_info[i].function_name))
-						strncpy(name, fixUnderscores(name), name_max);
-
-					if (strcmp(name, correct_frame_info[i].function_name) != 0) {
-						if (correct_frame_info[i].optional) {
-							j--;
-							continue;
-						}
-						logerror("**Failed** test %d (%s)\n", test_num, test_name);
-						logerror("    Stack frame #%d refers to function %s, should be %s\n", j+1, name, correct_frame_info[i].function_name);
-						failed = true;
-						break;
-					}
-				}
-			}
-		}
-	}
+    // check if each call in correct_frame_info has the same amount of calls in stack
+    vector<frameInfo_t> expected_stack(correct_frame_info, correct_frame_info+num_correct_names);
+    if(!hasSameAmountOfCalls(stack, expected_stack))
+    {
+        logerror("**Failed** test %d (%s)\n", test_num, test_name);
+        logerror("    Possibly duplicate (case of wrong recursion) in collected stack. ");
+        failed = true;
+    }
```

Comment:
Would it be good to also add a check that the PC and FP addresses in the computed stack are in strictly decreasing order? That could be done using std::is_sorted.

Reply:
The PC does not follow strictly decreasing order, as we don't know where the functions will be placed in memory, neither the base address of different libraries. So kill() could have a higher address compared to any test_stack_*_mutatee() or test_stack_*_func*(), while it is usually the last frame in these tests. But the FP for sure.